crates/aptos-crypto/src/webauthn/webauthn_p256_sigs.rs

```diff
@@ -0,0 +1,238 @@
+        &self,
+        message: &[u8],
+        public_key: &Self::VerifyingKeyMaterial,
+    ) -> anyhow::Result<()> {
+        let is_equal = self.verify_expected_challenge_from_message_matches_actual(
+            message,
+            self.actual_challenge.as_ref(),
+        );
+
+        // Check if expected challenge and actual challenge match. If there's no match, throw error
+        if !is_equal {
+            return Err(anyhow!(
+                "Error: WebAuthn expected challenge did not match actual challenge"
+            ));
+        }
+
+        // Verify signature against verification data
+        public_key
+            .0
+             .0
+            .verify(self.verification_data.as_slice(), &self.p256_signature.0)
+            .map_err(|e| anyhow!("{}", e))
+            .and(Ok(()))
+    }
+}
+
+//////////////////////
+// Signature Traits //
+//////////////////////
+
+impl Signature for WebAuthnP256Signature {
+    type VerifyingKeyMaterial = WebAuthnP256PublicKey;
+    type SigningKeyMaterial = WebAuthnP256PrivateKey;
+
+    /// Verifies that the provided signature is valid for the provided message, going beyond the
+    /// [NIST SP 800-186](https://csrc.nist.gov/publications/detail/sp/800-186/final) specification, to prevent scalar malleability as done in [BIP146](https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki).
+    fn verify<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+        public_key: &WebAuthnP256PublicKey,
+    ) -> anyhow::Result<()> {
+        Self::verify_arbitrary_msg(self, &signing_message(message)?, public_key)
+    }
+
+    /// Checks that `self` is valid for an arbitrary &[u8] `message` using `public_key`.
+    /// Outside of this crate, this particular function should only be used for native signature
+    /// verification in Move.
+    ///
+    /// This function will check both the signature and `public_key` for small subgroup attacks.
+    fn verify_arbitrary_msg(
+        &self,
+        message: &[u8],
+        public_key: &WebAuthnP256PublicKey,
+    ) -> anyhow::Result<()> {
+        self.verify_arbitrary_challenge(message, public_key)
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+
+impl Length for WebAuthnP256Signature {
+    fn length(&self) -> usize {
+        P256Signature::LENGTH
+    }
+}
+
+impl ValidCryptoMaterial for WebAuthnP256Signature {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+
+impl std::hash::Hash for WebAuthnP256Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_signature = self.to_bytes();
+        state.write(&encoded_signature);
+    }
+}
+
+impl TryFrom<&[u8]> for WebAuthnP256Signature {
+    type Error = CryptoMaterialError;
+
+    fn try_from(bytes: &[u8]) -> std::result::Result<WebAuthnP256Signature, CryptoMaterialError> {
+        WebAuthnP256Signature::check_s_malleability(bytes)?;
```

Comment:
Ideally we don't call `check_s_malleability` directly in this crate, it just makes it easier to get things wrong. Can our P256 `try_from` function be called somewhere here instead?

Reply:
I tried  implementing what you suggested, lmk what you think: https://github.com/aptos-labs/aptos-core/pull/9870/files#diff-1700c1053b859122e4f3074736abf6862f152f168356936df41925e8f6e3d559R201

Reply:
Note to self, after discussing with Michael, will change it to

```rust
impl TryFrom<&[u8]> for WebAuthnP256Signature {
    type Error = CryptoMaterialError;

    fn try_from(bytes: &[u8]) -> std::result::Result<WebAuthnP256Signature, CryptoMaterialError> {
        let signature = WebAuthnP256Signature::from_bytes_unchecked(bytes);
        // P256Signature::try_from(
        //     signature
        //         .clone()
        //        .unwrap()
        //      .p256_signature
        //        .to_bytes()
        //        .as_ref(),
        //)?;
        //signature
        P256Signature::check_s_malleability(signature.p256_signature.to_bytes())
    }
}

```